// utapi/dist/handlers/metricsHandlers.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.UsersHandler = exports.ServiceHandler = exports.AccountsHandler = exports.BucketsHandler = undefined;

var _ListMetrics = require('../lib/ListMetrics');

var _ListMetrics2 = _interopRequireDefault(_ListMetrics);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function listRecentMetrics(utapiRequest, metric, service, cb) {
    const log = utapiRequest.getLog();
    log.debug('handling list recent metrics request', {
        method: 'listRecentMetrics'
    });
    const Metric = new _ListMetrics2.default(metric, service);
    return Metric.getRecentTypesMetrics(utapiRequest, cb);
}

/**
 * @class BucketsHandler
 * Handles Buckets resource actions
 */

class BucketsHandler {
    /**
    * List metrics for the given list of buckets
    * @param {UtapiRequest} utapiRequest - UtapiRequest instance
    * @param {string} service - the service component (e.g., 's3')
    * @param {callback} cb - callback
    * @return {undefined}
    */
    static listMetrics(utapiRequest, service, cb) {
        const log = utapiRequest.getLog();
        log.debug('handling list metrics request', {
            method: 'BucketsHandler.listMetrics'
        });
        const Buckets = new _ListMetrics2.default('buckets', service);
        return Buckets.getTypesMetrics(utapiRequest, cb);
    }

    /**
    * List metrics starting from the second most recent fifteen minute timestamp
    * for the given list of buckets
    * @param {UtapiRequest} utapiRequest - UtapiRequest instance
    * @param {string} service - the service component (e.g., 's3')
    * @param {callback} cb - callback
    * @return {undefined}
    */
    static listRecentMetrics(utapiRequest, service, cb) {
        return listRecentMetrics(utapiRequest, 'buckets', service, cb);
    }
}

exports.BucketsHandler = BucketsHandler; /**
                                          * @class AccountsHandler
                                          * Handles Accounts resource actions
                                          */

class AccountsHandler {
    /**
    * List metrics for the given list of accounts
    * @param {UtapiRequest} utapiRequest - UtapiRequest instance
    * @param {string} service - the service component (e.g., 's3')
    * @param {callback} cb - callback
    * @return {undefined}
    */
    static listMetrics(utapiRequest, service, cb) {
        const log = utapiRequest.getLog();
        log.debug('handling list metrics request', {
            method: 'AccountsHandler.listMetrics'
        });
        const Accounts = new _ListMetrics2.default('accounts', service);
        return Accounts.getTypesMetrics(utapiRequest, cb);
    }

    /**
    * List metrics starting from the second most recent fifteen minute timestamp
    * for the given list of accounts
    * @param {UtapiRequest} utapiRequest - UtapiRequest instance
    * @param {string} service - the service component (e.g., 's3')
    * @param {callback} cb - callback
    * @return {undefined}
    */
    static listRecentMetrics(utapiRequest, service, cb) {
        return listRecentMetrics(utapiRequest, 'accounts', service, cb);
    }
}

exports.AccountsHandler = AccountsHandler; /**
                                            * @class ServiceHandler
                                            * Handles Services resource actions
                                            */

class ServiceHandler {
    /**
    * List metrics for the given list of services
    * @param {UtapiRequest} utapiRequest - UtapiRequest instance
    * @param {string} service - the service component (e.g., 's3')
    * @param {callback} cb - callback
    * @return {undefined}
    */
    static listMetrics(utapiRequest, service, cb) {
        const log = utapiRequest.getLog();
        log.debug('handling list metrics request', {
            method: 'ServiceHandler.listMetrics'
        });
        const Service = new _ListMetrics2.default('service', service);
        return Service.getTypesMetrics(utapiRequest, cb);
    }

    /**
    * List metrics starting from the second most recent fifteen minute timestamp
    * for the given list of services
    * @param {UtapiRequest} utapiRequest - UtapiRequest instance
    * @param {string} service - the service component (e.g., 's3')
    * @param {callback} cb - callback
    * @return {undefined}
    */
    static listRecentMetrics(utapiRequest, service, cb) {
        return listRecentMetrics(utapiRequest, 'service', service, cb);
    }
}

exports.ServiceHandler = ServiceHandler; /**
                                          * @class UsersHandler
                                          * Handles Users resource actions
                                          */

class UsersHandler {
    /**
    * List metrics for the given list of users
    * @param {UtapiRequest} utapiRequest - UtapiRequest instance
    * @param {string} service - the service component (e.g., 's3')
    * @param {callback} cb - callback
    * @return {undefined}
    */
    static listMetrics(utapiRequest, service, cb) {
        const log = utapiRequest.getLog();
        log.debug('handling list metrics request', {
            method: 'UsersHandler.listMetrics'
        });
        const Users = new _ListMetrics2.default('users', service);
        return Users.getTypesMetrics(utapiRequest, cb);
    }

    /**
    * List metrics starting from the second most recent fifteen minute timestamp
    * for the given list of users
    * @param {UtapiRequest} utapiRequest - UtapiRequest instance
    * @param {string} service - the service component (e.g., 's3')
    * @param {callback} cb - callback
    * @return {undefined}
    */
    static listRecentMetrics(utapiRequest, service, cb) {
        return listRecentMetrics(utapiRequest, 'users', service, cb);
    }
}
exports.UsersHandler = UsersHandler;
